File: VB/CustomizeNewActionItemsListExample.Module/Controllers/CustomizeNewActionItemsListController.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports DevExpress.ExpressApp
Imports DevExpress.Persistent.BaseImpl
Imports DevExpress.ExpressApp.SystemModule

Namespace CustomizeNewActionItemsListExample.Module.Controllers
    Public Class CustomizeNewActionItemsListController
        Inherits ObjectViewController(Of ObjectView, Task)

        Protected Overrides Sub OnActivated()
            MyBase.OnActivated()
            Dim controller As NewObjectViewController = Frame.GetController(Of NewObjectViewController)()
            If controller IsNot Nothing Then
                AddHandler controller.CollectCreatableItemTypes, AddressOf NewObjectViewController_CollectCreatableItemTypes
                AddHandler controller.CollectDescendantTypes, AddressOf NewObjectViewController_CollectDescendantTypes
                If controller.Active Then
                    controller.UpdateNewObjectAction()
                End If
            End If
        End Sub
        Private Sub NewObjectViewController_CollectDescendantTypes(ByVal sender As Object, ByVal e As CollectTypesEventArgs)
            CustomizeList(e.Types)
        End Sub
        Private Sub NewObjectViewController_CollectCreatableItemTypes(ByVal sender As Object, ByVal e As CollectTypesEventArgs)
            CustomizeList(e.Types)
        End Sub
        Private Sub CustomizeList(ByVal types As ICollection(Of Type))
            Dim unusableTypes As New List(Of Type)()
            For Each item As Type In types
                If item Is GetType(Task) Then
                    unusableTypes.Add(item)
                End If
            Next item
            For Each item As Type In unusableTypes
                types.Remove(item)
            Next item
        End Sub
        Protected Overrides Sub OnDeactivated()
            Dim controller As NewObjectViewController = Frame.GetController(Of NewObjectViewController)()
            If controller IsNot Nothing Then
                RemoveHandler controller.CollectCreatableItemTypes, AddressOf NewObjectViewController_CollectCreatableItemTypes
                RemoveHandler controller.CollectDescendantTypes, AddressOf NewObjectViewController_CollectDescendantTypes
            End If
            MyBase.OnDeactivated()
        End Sub
    End Class
End Namespace

File: CS/CustomizeNewActionItemsListExample.Module/Controllers/CustomizeNewActionItemsListController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.ExpressApp;
using DevExpress.Persistent.BaseImpl;
using DevExpress.ExpressApp.SystemModule;

namespace CustomizeNewActionItemsListExample.Module.Controllers {
    public class CustomizeNewActionItemsListController : ObjectViewController<ObjectView, Task> {
        protected override void OnActivated() {
            base.OnActivated();
            NewObjectViewController controller = Frame.GetController<NewObjectViewController>();
            if (controller != null) {
                controller.CollectCreatableItemTypes += NewObjectViewController_CollectCreatableItemTypes;
                controller.CollectDescendantTypes += NewObjectViewController_CollectDescendantTypes;
                if (controller.Active) {
                    controller.UpdateNewObjectAction();
                }
            }
        }
        private void NewObjectViewController_CollectDescendantTypes(object sender, CollectTypesEventArgs e) {
            CustomizeList(e.Types);
        }
        private void NewObjectViewController_CollectCreatableItemTypes(object sender, CollectTypesEventArgs e) {
            CustomizeList(e.Types);
        }
        private void CustomizeList(ICollection<Type> types) {
            List<Type> unusableTypes = new List<Type>();
            foreach (Type item in types) {
                if (item == typeof(Task)) {
                    unusableTypes.Add(item);
                }
            }
            foreach (Type item in unusableTypes) {
                types.Remove(item);
            }
        }
        protected override void OnDeactivated() {
            NewObjectViewController controller = Frame.GetController<NewObjectViewController>();
            if (controller != null) {
                controller.CollectCreatableItemTypes -= NewObjectViewController_CollectCreatableItemTypes;
                controller.CollectDescendantTypes -= NewObjectViewController_CollectDescendantTypes;
            }
            base.OnDeactivated();   
        }
    }
 }

